packages/field-base/src/char-length-mixin.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * @license
+ * Copyright (c) 2021 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { dedupingMixin } from '@polymer/polymer/lib/utils/mixin.js';
+import { ForwardInputPropsMixin } from './forward-input-props-mixin.js';
+
+const CharLengthMixinImplementation = (superclass) =>
+  class CharLengthMixinClass extends ForwardInputPropsMixin(superclass) {
+    static get properties() {
+      return {
+        /**
+         * Maximum number of characters (in Unicode code points) that the user can enter.
+         */
+        maxlength: {
+          type: Number
+        },
+
+        /**
+         * Minimum number of characters (in Unicode code points) that the user can enter.
+         */
+        minlength: {
+          type: Number
+        }
+      };
+    }
+
+    static get forwardProps() {
+      return [...super.forwardProps, 'maxlength', 'minlength'];
+    }
+
+    static get constraints() {
+      return [...super.constraints, 'maxlength', 'minlength'];
+    }
+
+    /**
+     * Override an observer from `ValidateMixin` to add `minlength` and `maxlength` constraints.
+     * @param {boolean | undefined} required
+     * @param {number | undefined} minlength
+     * @param {number | undefined} maxlength
+     * @protected
+     * @override
+     */
+    _constraintsChanged(required, minlength, maxlength) {
+      // Prevent marking field as invalid when setting required state
+      // or any other constraint before a user has entered the value.
+      if (!this.invalid) {
+        return;
+      }
+
+      if (!required && !minlength && !maxlength) {
+        this.invalid = false;
+      } else {
+        this.validate();
+      }
+    }
+
+    /**
+     * Returns true if the current input value satisfies all constraints (if any).
+     * @return {boolean}
+     */
+    checkValidity() {
+      if (this.required || this.maxlength || this.minlength) {
+        return this.inputElement ? this.inputElement.checkValidity() : undefined;
```

Comment:
Is this method expected to return `undefined`? The JS Doc doesn't specify this as expected output.

The same question goes for some other mixins, such as:

- [PatternMixin](https://github.com/vaadin/web-components/blob/bb03baecd58dbaf13fc6fabd0feddf0178b3eb55/packages/field-base/src/pattern-mixin.js#L104)
- [TextFieldMixin](https://github.com/vaadin/web-components/blob/bb03baecd58dbaf13fc6fabd0feddf0178b3eb55/packages/field-base/src/text-field-mixin.js#L73)
- [InputFieldMixin](https://github.com/vaadin/web-components/blob/bb03baecd58dbaf13fc6fabd0feddf0178b3eb55/packages/field-base/src/input-field-mixin.js#L132)

Reply:
Good catch. Fixed for all the mixins that implement `checkValidity` and added tests for this case.